### PR TITLE
Early stopping and other parameters for learning rate scheduler

### LINKED ISSUE #183: Support for other schedulers and early stopping
I was hoping to use a different learning rate scheduler than the default implementation (reduce by 0.8 after a validation loss plateau with patience of 16 epochs). Unfortunately it seems like adding this to the codebase might be rather hairy. My current idea is to add the scheduler information (except for the optimizer) in the config file, then pass a `_Proxy` object with that information to `TrainModel`. Then the optimizer can be bound to the proxy and a scheduler object instance can be created when `_init_train` is called. Does this seem like an appropriate approach, or is there a better design solution?

It would also be useful to add support for early stopping. As I see it, this seems fairly straightforward to me -- all that is needed is to add information in the configs about which metric should be used for early stopping and how much patience. That information can be bound to `TrainModel` and utilized within `train_and_validate`.

### PR DESCRIPTION
**Pull request by rfriedman22**
#### Reference Issues/PRs

Fixes #183 

#### What does this implement/fix? Explain your changes.
1. Added support for other parameters for `ReduceLROnPlateau`. The `TrainModel` object has a new argument `scheduler_kwargs` that defaults to the pre-existing default parameters. User can specify other parameters as a dictionary, including through YAML config files. The kwargs get passed to the `_init_train` method when the scheduler is created. Support for other types of schedulers was not added.
2. Added support for early stopping. The `TrainModel` object has a new optional argument `stopping_criteria`. If specified, it is a list that must correspond to the metric and patience for stopping criteria. When `validate` is called, it checks to see the last time the specified metric improves on the validation set. If the metric has not improved in the specified patience period, then the logger reports a debug message that early stopping has been reached and the for loop in `train_and_validate` is terminated.

#### What testing did you do to verify the changes in this PR?
1. Created `TrainModel` objects from YAML config files that contain the new arguments. Verified the `scheduler` is assigned parameters specified in the config. Verified the early stopping criteria gets set as internal variables.
2. I had previously implemented the early stopping code as a hack in my own scripts and verified that the criteria can be reached and stops training early. Code implemented here is virtually identical to what I had used in my scripts.

### CODE REVIEW

**Comment by kathyxchen** [comment]
Thanks so much for this PR @rfriedman22 ! Is there an example you can add to `config_examples` or just a code snippet you could leave here for us to add to the website for this new change? LGTM (:

**Comment by rfriedman22** [comment]
Sorry for the delay @kathyxchen -- it's been a busy time lately. I've added an example config to this PR that demonstrates the new features :)
